libs/dyn/visit_set.go

```diff
@@ -14,11 +20,59 @@ func Set(v Value, path string, nv Value) (Value, error) {
 // If successful, it returns the new value with all intermediate values copied and updated.
 // If the path doesn't exist, it returns InvalidValue and an error.
 func SetByPath(v Value, p Path, nv Value) (Value, error) {
-	return visit(v, EmptyPath, p, visitOptions{
-		fn: func(_ Path, _ Value) (Value, error) {
-			// Return the incoming value to set it.
-			return nv, nil
+	lp := len(p)
+	if lp == 0 {
+		return nv, nil
+	}
+
+	parent := p[:lp-1]
+	component := p[lp-1]
+
+	return visit(v, EmptyPath, parent, visitOptions{
+		fn: func(prefix Path, v Value) (Value, error) {
+			path := prefix.Append(component)
+
+			switch {
+			case component.isKey():
+				// Expect a map to be set if this is a key.
+				m, ok := v.AsMap()
+				if !ok {
+					return InvalidValue, fmt.Errorf("expected a map to index %q, found %s", path, v.Kind())
+				}
+
+				// Return an updated map value.
+				m = maps.Clone(m)
+				m[component.key] = nv
+				return Value{
+					v: m,
+					k: KindMap,
+					l: v.l,
+				}, nil
+
+			case component.isIndex():
+				// Expect a sequence to be set if this is an index.
+				s, ok := v.AsSequence()
+				if !ok {
+					return InvalidValue, fmt.Errorf("expected a sequence to index %q, found %s", path, v.Kind())
+				}
+
+				// Lookup current value in the sequence.
+				if component.index < 0 || component.index >= len(s) {
+					return InvalidValue, indexOutOfBoundsError{prefix}
```

Comment:
What would be the idiomatic way to append to a sequence using `dyn`?

Reply:
For an existing sequence, this would be to call `dyn.Map` on it and the modify the sequence yourself.